xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -0,0 +1,299 @@
+                XdsClient.XdsChannelFactory.getInstance(),
+                node,
+                syncContext,
+                timeService,
+                new ExponentialBackoffPolicy.Provider(),
+                GrpcUtil.STOPWATCH_SUPPLIER);
+          }
+        };
+    xdsClientPool = new XdsClient.RefCountedXdsClientObjectPool(xdsClientFactory);
+    xdsClient = xdsClientPool.getObject();
+    setClientAndWatcher(port);
+  }
+
+  @VisibleForTesting
+  XdsClientWrapperForServerSds(int port, XdsClient xdsClient) {
+    this.logId = InternalLogId.allocate("GrpcServer", Integer.toString(port));
+    this.xdsClient = xdsClient;
+    setClientAndWatcher(port);
+  }
+
+  private void setClientAndWatcher(int port) {
+    xdsClient.watchListenerData(
+        port,
+        new XdsClient.ListenerWatcher() {
+          @Override
+          public void onListenerChanged(XdsClient.ListenerUpdate update) {
+            logger.log(
+                Level.INFO,
+                "Setting myListener from ConfigUpdate listener :{0}",
+                update.getListener().toString());
+            curListener = update.getListener();
+          }
+
+          @Override
+          public void onError(Status error) {
+            // In order to distinguish between IO error and resource not found, set curListener
+            // to null in case of NOT_FOUND
+            if (error.getCode().equals(Status.Code.NOT_FOUND)) {
+              curListener = null;
+            }
+            // TODO(sanjaypujare): Implement logic for other cases based on final design.
+            logger.log(Level.SEVERE, "ListenerWatcher in XdsClientWrapperForServerSds:{0}", error);
+          }
+        });
+  }
+
+  /**
+   * Locates the best matching FilterChain to the channel from the current listener and returns the
+   * DownstreamTlsContext from that FilterChain.
+   */
+  public DownstreamTlsContext getDownstreamTlsContext(Channel channel) {
+    if (curListener != null && channel != null) {
+      List<FilterChain> filterChains = curListener.getFilterChains();
+      int highestScore = -1;
+      FilterChain bestMatch = null;
+      for (FilterChain filterChain : filterChains) {
+        int curScore = getFilterChainMatchScore(filterChain.getFilterChainMatch(), channel);
+        if (curScore > 0 && curScore > highestScore) {
+          bestMatch = filterChain;
+          highestScore = curScore;
+        }
+      }
+      if (bestMatch != null) {
+        return bestMatch.getDownstreamTlsContext();
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Computes a score for a match of filterChainMatch with channel.
+   *
+   * <p>-1 => mismatch (of port or IP address etc)
+   *
+   * <p>0 => unimplemented (types or logic etc)
+   *
+   * <p>1 => filterChainMatch is null, so nothing to match.
+   *
+   * <p>value > 2 indicates some kind of IP address match.
+   */
+  private static int getFilterChainMatchScore(FilterChainMatch filterChainMatch, Channel channel) {
```

Comment:
In the original filterChainMatch (from proto) there are other fields like the remote address,  which will ideally need to be matched with the remote address, remote port in the Channel. Trying to implement all the logic now seems like an overkill now since most filterChainMatch don't use it but it makes sense to pass Channel instead of separate fields from the Channel. Agree?

Reply:
it is a private method so we can always change if we need more fields. also, if you use comparator it may look very different.

Reply:
Yes, with comparator it will use localAddress now and Channel later.